src/coupled_driver.cpp

```diff
@@ -267,216 +290,274 @@ bool CoupledDriver::is_converged()
+
+  for (const auto& heat_rank : heat_ranks_) {
+    comm_.send_and_recv(cells_recv, neutronics_root_, cells_, heat_rank);
+    neutronics.comm_.broadcast(cells_recv);
+
+    comm_.send_and_recv(cell_volumes_recv, neutronics_root_, cell_volumes_, heat_rank);
+    neutronics.comm_.broadcast(cell_volumes_recv);
+
+    comm_.send_and_recv(
+      cell_densities_recv, neutronics_root_, cell_densities_, heat_rank);
+    neutronics.comm_.broadcast(cell_densities_recv);
+
+    comm_.send_and_recv(
+      cell_fluid_mask_recv, neutronics_root_, cell_fluid_mask_, heat_rank);
+    neutronics.comm_.broadcast(cell_fluid_mask_recv);
+
+    if (neutronics.active()) {
+      for (gsl::index i = 0; i < cells_recv.size(); ++i) {
+        if (cell_fluid_mask_recv.at(i) == 1) {
+          auto rho = cell_densities_recv.at(i);
+          auto V = cell_volumes_recv.at(i);
+          cell_V[cells_recv.at(i)] += V;
+          rho_dot_V[cells_recv.at(i)] += rho * V;
         }
-        // Set density for cell instance
-        average_density /= total_vol;
-        Ensures(average_density > 0.0);
-        neutronics.set_density(cell, average_density);
       }
     }
   }
+  for (const auto& kv : rho_dot_V) {
+    neutronics.set_density(kv.first, kv.second / cell_V.at(kv.first));
+  }
 }
 
 void CoupledDriver::init_mappings()
 {
   comm_.message("Initializing mappings");
-
   const auto& heat = this->get_heat_driver();
   auto& neutronics = this->get_neutronics_driver();
 
-  // Get centroids from heat driver and send to all neutronics procs
-  auto elem_centroids = heat.centroids();
-  comm_.send_and_recv(elem_centroids, neutronics_root_, heat_root_);
-  neutronics.comm_.broadcast(elem_centroids);
-
-  if (neutronics.active()) {
-    // Get cell handle corresponding to each element centroid
-    elem_to_cell_ = neutronics.find(elem_centroids);
-
-    // Create a vector of elements for each neutronics cell
-    for (int32_t elem = 0; elem < elem_to_cell_.size(); ++elem) {
-      auto cell = elem_to_cell_[elem];
-      cell_to_elems_[cell].push_back(elem);
+  // Send and recv buffers
+  std::vector<Position> centroids_send, centroids_recv;
+  decltype(elem_to_cell_) elem_to_cell_send;
+
+  for (const auto& heat_rank : heat_ranks_) {
+    // For the given heat rank, the neutronics root discovers the mapping of
+    // local elem ID --> global cell handle.
+    // * IMPORTANT: OpenmcDriver::find adds the cell instances it discovers to
+    //   the OpenmcDriver::cells_ array of the calling rank only.  However, every
+    //   neutronics rank needs the full array of cells_ for Openmc::create_tallies.
+    //   Hence, we broadcast the centroids to all neutronics ranks and then
+    //   call neutronics.find on each neutronics rank.
+    if (comm_.rank == heat_rank) {
+      centroids_send = heat.centroid_local();
+    }
+    this->comm_.send_and_recv(
+      centroids_recv, neutronics_root_, centroids_send, heat_rank);
+    neutronics.comm_.broadcast(centroids_recv);
```

Comment:
This part is a little weird.  `OpenmcDriver::find()`, directly below, has the essential side-effect of populating the `OpenmcDriver::cells_` array.  Furthermore, `OpenmcDriver::create_tallies` needs to be called from every neutronics rank with the same `cells_` array.  Hence, we need to broadcast `centroids_recv` to all the neutronics ranks so that every neutroincs ranks can call `OpenmcDriver::find` and populate `OpenmcDriver::cells_` the same way.